### DISCUSSION TITLE
F411 Mapping RX2/RX2 as Servo Output

### DISCUSSION THREAD

**Discussion by lydianb79** [iNavFlight/inav#10480] [General]
Hi,

I am currently trying to get RX2/TX2 on an F411 to work as Servo Output.
After spending hours on the Datasheet and compiling x different Variants, I am unable to get output on the Servo.
I am using IVAV 7.1.2 and I am aware that F411 will not be compatible anymore with INAV8 and it's probably not the best target to do this.
I am currently in a learning Phase and try to understand how things work.
If anyone could point me to the correct timer, channels and pins, that would be awesome so that I can learn why it did not work for me so far.
Thanks!

**Comment by sensei-hacker** [comment]
You haven't provided the diff of what you've tried, so you're asking us to tell you what's wrong with some code without seeing the code. Meaning the best we could do is make wild guesses. Here's a wild guess:
You forgot to remove usart2, so you're trying to use it as both a PWM and a uart at the same time. That guess is probably wrong, but without seeing your code a wild guess is the best we can do.

You didn't mention which flight controller you're talking about. So UART2 could be on PA2 and PA3, or it could be on PD5 and PD6. You might need to use timer 2, or 5, or 9 if it's using the PA pins, depending on how each timer is currently being used.  Or there may be no at all timers that can be connected (the PD pins).

Pretty much you're asking people to debug your code without seeing it, and determine why it doesn't match your hardware - without telling us which hardware. :)

**Reply by lydianb79** [reply]
Thanks!
Sorry for not being specific enough. As I said I am in a Learning Process. The target is BETAFPVF411
Below the diffs from target.h and target.c
[diff.txt](https://github.com/user-attachments/files/17893807/diff.txt)

I tried with all Timers and Channels from the Datasheet for PA2 / PA3

<img width="741" alt="Bildschirmfoto 2024-11-24 um 16 26 24" src="https://github.com/user-attachments/assets/8aa13aef-cff3-49fd-8a5c-17cbd6d4c9aa">

**Comment by sensei-hacker** [comment]
I notice your "change" to MAX_PWM_OUTPUT_PORTS is commented out, so that hasn't been updated.

+ // #define MAX_PWM_OUTPUT_PORTS       5
  #define MAX_PWM_OUTPUT_PORTS       4

**Reply by lydianb79** [reply]
I changed it from 5 to 4 since I commented out

> // #define USE_UART2
> // #define UART2_TX_PIN            PA2
> // #define UART2_RX_PIN            PA3

**Reply by sensei-hacker** [reply]
> I changed it from 5 to 4 since I commented out
> 
> > // #define USE_UART2
> > // #define UART2_TX_PIN            PA2
> > // #define UART2_RX_PIN            PA3

You are trying to INCREASE the number of PWM outputs, aren't you?

**Reply by lydianb79** [reply]
hmmm, actually yes but as far as I could understand from here:
https://github.com/BossHobby/inav-wing-aio/releases/tag/7.1.2-whoop-wing

This was commented out on the already working targets and the Number of Ports reduced by 1. Is this not what you see in INAV in the Ports Tab? UART2 should not show up there since it can't be used for PWM and Port at the same time?

**Comment by sensei-hacker** [comment]
You are wanting to decrease SERIAL_PORT_COUNT in order to increase MAX_PWM_OUTPUT_PORTS.

**Reply by lydianb79** [reply]
You are right, I am on the wrong track here. I must have missunderstood something. Thanks a lot for your Input, highly appreciated!
I mixed up SERIAL_PORT_COUNT and MAX_PWM_OUTPUT_PORTS

Not sure if that was correct when I did the last builds, I try again :-)

**Comment by lydianb79** [comment]
Solved, the Issue was that TIM2 was already used by a Soft Serial. Since I don't need it, I ditched it and voila!
Definitely learned a lot a in the last days!
